src/mon/OSDMonitor.cc

```diff
@@ -1437,7 +1437,8 @@ epoch_t OSDMonitor::get_min_last_epoch_clean() const
   // also scan osd epochs
   // don't trim past the oldest reported osd epoch
   for (auto& osd_epoch : osd_epochs) {
-    if (osd_epoch.second < floor) {
+    if (osd_epoch.second < floor &&
```

Comment:
guess it might be too early to trim epochs of a "down" OSD. how about being more tolerant, and only trimming an "out" OSD?

also, can we just remove the newly "out" osd's epoch from `osd_epochs` in `OSDMonitor::update_from_paxos()`, by checking `inc.new_weight`?  see https://github.com/ceph/ceph/blob/master/src/mon/OSDMonitor.cc#L433

but yes, probably i am too conservative. as the chance of all OSD being marked down for a PG at the same time is rare. in the case, it'd be easier and more consistent to trim the `osd_epochs` just by looking at `inc.new_state`.

Reply:
What if we have some massive outage and a significant fraction of OSDs are marked down and out -- will the mon then trim too many osdmaps? Could that cause a bigger problem?

Reply:
@dvanders I don't think so. We're getting the lowest epoch at which point the pools were clean, and we're not trimming past that. The objective here is to just remove maps (at which point the pools were clean) without accounting for out osds, which would otherwise keep us from trim past their last reported epoch.

Reply:
> guess it might be too early to trim epochs of a "down" OSD. how about being more tolerant, and only trimming an "out" OSD?

I don't think we get much of a benefit from accounting only for out osds rather than down, aside from time itself. We're not removing past a given state of cleanliness. As far as I understand, if we end up with all copies of a PG down, then that pool is not clean and we can't clean past that epoch. Now, maybe it's worth going with `out` instead of `down` simply because it's common for users to set `noout` when doing maintenance (upgrades, etc), and we may want to reduce the load on the cluster (including the monitor), and avoid a significant trim at that point in time. But, to be honest, even there I'm not sure we get much of a benefit.

> also, can we just remove the newly "out" osd's epoch from `osd_epochs` in `OSDMonitor::update_from_paxos()`, by checking `inc.new_weight`? see https://github.com/ceph/ceph/blob/master/src/mon/OSDMonitor.cc#L433

I'm not sure about this. We already remove the state from `osd_epochs` when an osd ceases to exist; I don't think we should subvert that notion by also clearing it when it goes out/down/whatever. Until the osd is purged, the osd still exists and, at least conceptually, should be kept in the `osd_epochs` map (although I'm pretty sure we're keeping destroyed osds when we probably shouldn't).

> but yes, probably i am too conservative. as the chance of all OSD being marked down for a PG at the same time is rare. in the case, it'd be easier and more consistent to trim the `osd_epochs` just by looking at `inc.new_state`.

Reply:
makes sense to me.